### PR TITLE
Move template compiler creation to a method on the addon

### PR DESCRIPTION
**Pull request by chriseppstein**
This is so it can be overridden by plugins (e.g. CSS Blocks) that need access to the input and output trees.

Relevant to this patch:

* Here's where I [override this method](https://github.com/linkedin/css-blocks/blob/ember-rewrite/packages/%40css-blocks/ember/src/index.ts#L191-L197).
* Here's where I [inherit from the template compiler](https://github.com/linkedin/css-blocks/blob/ember-rewrite/packages/%40css-blocks/ember/src/index.ts#L42-L55).

Also, I added some typescript types for the template compiler. I can split that off into a separate PR if you want.

Tests? I can add one but I'm not sure exactly what you'd like for me to test. It seems like maybe just a code comment is sufficient.

### CODE REVIEW

File: lib/ember-addon-main.js [hunk 1]
```diff
@@ -56,6 +56,11 @@ module.exports = {
     return this._cachedShouldColocateTemplates;
   },
 
+  getTemplateCompiler(inputTree, htmlbarsOptions) {
```
Comment (stefanpenner) [inline lib/ember-addon-main.js:59]: If this is new public API we may want to:

* add details to the readme
* test that this "hook" can be used as expected

Comment (chriseppstein) [reply lib/ember-addon-main.js:60]: * I've added details to the README.
* I didn't add a test case because I'm not sure how. There's no tests for that unit of code other than the one that runs the addon on the dummy application.

**Comment by chriseppstein** [comment]
I've upgraded to broccoli-persistent-filter 3.0 which allows us to pick up the proper typescript definitions for that plugin now and use it here.

File: README.md [hunk 2]
```diff
@@ -160,3 +160,47 @@ var templateTree = new HtmlbarsCompiler('app/templates', {
...
+    if (this.parent.findOwnAddonByName) {
+      return this.parent.findOwnAddonByName(name);
+    } else {
+      return this.project.findAddonByName(name);
+    }
+  },
+
+  included() {
+    this._super.included.apply(this, arguments);
+    const htmlBarsAddon = this.findSiblingAddon('ember-cli-htmlbars');
```
Comment (rwjblue) [inline README.md:195]: I'd prefer this (and to avoid the `findSiblingAddon` helper method):

```js
const htmlbarsAddon = this.parent.addons.find(a => a.name === 'ember-cli-htmlbars');
```

File: lib/template-compiler-plugin.d.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import Filter = require('broccoli-persistent-filter');
```
Comment (rwjblue) [inline lib/template-compiler-plugin.d.ts:1]: I do not want to ship ambient module declarations for modules that we actually provide (the ambient module declarations that we have at the moment are strictly for the fabricated `hbs` entry point). I'm happy to entertain a refactor to leverage TS itself as a separate PR but that is unrelated to the work being done in this PR. When we do that, we'll have to nail down the proper SemVer strategy.

**Comment by rwjblue** [comment]
Landed the broccoli-persistent-filter bump in https://github.com/ember-cli/ember-cli-htmlbars/pull/557.

File: lib/ember-addon-main.js [hunk 4]
```diff
@@ -57,6 +57,11 @@ module.exports = {
...
+    const TemplateCompiler = require('./template-compiler-plugin');
+    return new TemplateCompiler(inputTree, htmlbarsOptions);
+  },
```
Comment (rwjblue) [inline lib/ember-addon-main.js:63]: I'd much prefer to provide a `transpileTree` API akin to the one in ember-cli-babel (see [documentation here](https://github.com/babel/ember-cli-babel#additional-trees)) which could be used from your own addon.

In that setup, you would unregister `ember-cli-htmlbars` and register your own addon as the preprocessor. Telling folks to monkey patch our internal methods seems pretty error prone to me (both for us, and for them) 🤔.

File: lib/template-compiler-plugin.d.ts (hunk 3 above)
Comment (chriseppstein) [reply lib/template-compiler-plugin.d.ts]: I removed this types file.

File: lib/ember-addon-main.js (hunk 4 above)
Comment (chriseppstein) [reply lib/ember-addon-main.js]: I've renamed the method to `transpileTree` and removed the README instructions telling people they should monkey patch this addon.
